### ISSUE TITLE
Question: How to debug blocks?

### ISSUE THREAD

**Issue by xxxserxxx** [greshake/i3status-rust#669]
Is there an env var to enable debugging? I don't see a command line option to enable debugging.

I have a toggle block with a `command-state` that, when executed on the command line, outputs "1" on STDOUT and a success exit code.  The toggle remains "off" in the bar.  Manually toggling it triggers something, but it flips back almost immediately. I'd like to debug this.  I don't see any output in `.xsession-errors` that looks like it comes from i3status-rust, although when I created a simple test that should always have been "true," whenever I toggle the widget I get a "1" in the `.xsession-errors` log.

What's the correct way of debugging widgets like this?

**Comment by ammgws** [comment]
There's no debug output at the moment.

What version are you using?
Can you give me a minimal reproducible example that I could test with?

**Comment by xxxserxxx** [comment]
`-V` says 0.13.0, but it's cloned and built from git; `rev-list` HEAD gives me 5e6e1ed4e47ca38b21cb5268251995ed23fd9708.

**Any** command produces the same behavior.  `ls`. `yes | head -1`.  Here's a version using "echo":
```
[[block]]
block = "toggle"
text = "Test"
command_state = "echo true"
command_on = "echo true"
command_off = "echo false"
interval = 2
```
It all behaves the same way.  A couple of weeks ago, I had something working, but I changed what I was toggling about a week ago and it stopped working; I thought it was my script, but now I can't get toggle to work at all.  With the above block, the toggle is off.  If I manually toggle it on, it turns itself off.
```

I'm sure it's something I'm doing, but I can't figure it out; hence the request for debugging.

**Comment by xxxserxxx** [comment]
I just pulled and compiled, and now i3bar is **really** unhappy, because i3status-rust is putting a "," at the end of every response.  I *think* that's what it's complaining about, anyway -- the `.xsession-errors` is filling up with these errors:
```
[i3bar] Could not parse JSON input (code = 2, message = parse error: after array element, I expect ',' or ']'): 
```

**Comment by xxxserxxx** [comment]
Reverting commit 88bfba7efbd2a4832857f337cbcf815e23ad7932 doesn't make Toggle work for me, but it stops the block from crashing i3bar.

**Comment by ammgws** [comment]
> because i3status-rust is putting a "," at the end of every response

It's meant to do that, so it's probably something else. I wonder if it's due to the output of your commands. Does it still give those errors with your `echo` example from above?

This is your example block running with i3status-rs git master:
https://i.imgur.com/Ci9IJLm.mp4
That's how you expect it to behave, right?

**Comment by ammgws** [comment]
The block crashing issue should be fixed with #674 

What happens if you try your test config with the below?
```toml
command_on = "exit 0"
command_off = "exit 1"
```

**Comment by xxxserxxx** [comment]
Haven't had a chance to process your comments yet, but am looking at it now.

**Comment by ammgws** [comment]
No worries. If it's still not fixed, if you could post your actual command then I could look into it further.

**Comment by xxxserxxx** [comment]
@ammgws I edited my comment, but you got here before I did.  Please see above.

FWIW, the only argument I'm running i3status-rs with is the location of the config file.

**Comment by ammgws** [comment]
> restarting i3, it comes up with the toggle off. If I manually toggle it, it flashes red and then goes back immediately to off.

If the toggle command exits with a non-zero status, the block will not be toggled.

The example config below should toggle:

> [[block]]
> block = "toggle"
> text = "Test"
> command_state = "echo true"
> command_on = "exit 0"
> command_off = "exit 0"
> interval = 2

If that works, that means your toggle command is exiting with a non-zero exit status. Can you check why that's happening?

**Comment by xxxserxxx** [comment]
Aha.  Ok, so I changed the on/off commands to echo to a file, and played around with various `command_state` commands and I think it's behaving -- after #674 I just needed to get back to a sane place.  Now the trivial case works; I still can't get the `command_state` to do what I'd expect, so maybe I'm still not understanding something.

What I have for a command state is:
```
command_state="timew get dom.active | grep 1"
```
The idea is that if TimeWarrior has an active tracker (`dom.active` == 1), the grep succeeds both in outputting text and exiting with `$?` == 0.  This, I can verify by running the command on the command line and checking the exit value.  However, the toggle *never* sees the _command_state_ as "on" -- I can toggle it manually, and it starts the command_on program (`ps` shows it running in the background), but the toggle switches itself back off after the interval and stays off.  If I start and stop the tracker on the command line, the toggle reacts (gratifyingly!) so the test is working -- I just can't figure out why it fails and stays off if toggled manually.  It'd be really helpful to be able to see what the toggle block thinks is going on.

However, now that I'm getting different behavior after the update, I need to poke and prod this some more; maybe I can get it working.

The original ticket request, which wasn't about my specific issue but rather to have some debugging ability, is still something I believe would be useful.  However, if that's a "won't fix" then I'll close this once I get my specific use case working.

**Comment by ammgws** [comment]
I installed TimeWarrior and ran `timew start test`. 
`"timew get dom.active | grep 1` prints "1" to stdout and exits cleanly.

Ran i3status-rust with this config:
```toml
[[block]]
block = "toggle"
text = "Test"
command_state = "timew get dom.active | grep 1"
command_on = "exit 0"
command_off = "exit 0"
interval = 5
```
Block is behaving as expected: https://imgur.com/a/2QqJWyx

**Comment by ammgws** [comment]
> The original ticket request, which wasn't about my specific issue but rather to have some debugging ability, is still something I believe would be useful. However, if that's a "won't fix" then I'll close this once I get my specific use case working.

I wonder what other blocks it might be useful for though.

**Comment by xxxserxxx** [comment]
`grep` was exiting non-zero if it didn't match, which was preventing `toggle` from doing anything.

So: for toggle to work properly:
1. `command_state` has to print something out to mean "on"
2. `command_state` has to print *nothing* out to mean "off" (both of these are documented)
3. `command_state` has to exit successfully, $? == 0 (this one is *not* documented)
4. The exit codes and output of `command_on` and `command_off` are irrelevant.  I'm not sure about this one, but it *appears* to be the case.

As for whether or not the ability to enable debugging to aid in ticket reports and generally getting things working -- well, *I* think it's useful, but it's not a core feature.

I'm going to close this, and add rule #3 to the wiki documentation.

**Comment by ammgws** [comment]
> command_state has to exit successfully, $? == 0 (this one is not documented)

Hmm, wonder if this is intentional or a bug... I will look into it.

Personally I think `command_state` should behave the same as the on/off commands: exit 0 for ON, otherwise for OFF. But it would be a breaking change.
